package/base-files/files/lib/functions/leds.sh

```diff
@@ -12,14 +12,26 @@ get_dt_led_path() {
 	echo "$ledpath"
 }
 
+get_dt_led_name() {
+	local dtname=$(basename "$1")
+	local dtparent=$(basename $(dirname $1))
+	local ledpath
+	local ledname
+
+	[ -n "$dtname" ] && ledpath=$(grep -rl '^OF_NAME='$dtname'$' /sys/devices/platform/$dtparent/leds/*/uevent)
+	[ -n "$ledpath" ] && ledname=$(echo $ledpath | sed 's#.*/\([^/]*\)/uevent#\1#')
+
+	[ -n "$ledname" ] && echo "$ledname" || echo "$dtname"
+}
+
 get_dt_led() {
 	local label
 	local ledpath=$(get_dt_led_path $1)
 
 	[ -n "$ledpath" ] && \
 		label=$(cat "$ledpath/label" 2>/dev/null) || \
 		label=$(cat "$ledpath/chan-name" 2>/dev/null) || \
-		label=$(basename "$ledpath")
```

Comment:
we can leave the basename there as a last resort. The led's common.yaml [in 56](https://elixir.bootlin.com/linux/v5.9-rc5/source/Documentation/devicetree/bindings/leds/common.yaml#L56) still states "If omitted, the label is taken from the node name [...]" and drivers like [leds-pwm.c](https://elixir.bootlin.com/linux/v5.9-rc5/source/drivers/leds/leds-pwm.c#L119) have not been converted. It's a barrel of fun.

Here's the output for the MR32 (which uses gpio-leds for and leds-pwm)

    MR32# ls -sys/class/leds
    amber:fault      blue             green            red              white:indicator

Reply:
... And since I wrote it anyway. There's another worse way to do it. This is the leds.sh portion. It uses a lookup table and needs to convert the color big-endian (because DT is big endian) value to a string.

    hex_be32_to_cpu() {
        [ "$(echo 10 | hexdump -v -n 2 -e '/2 "%x"')" = "3031" ] && {
                echo "${1:8:2}${1:6:2}${1:4:2}${1:2:2}${1:0:2}"
                return
        }
        echo "$@"
    }

    get_dt_color_by_id() {
        val=$(( 0x$(hex_be32_to_cpu $(hexdump -v -e '/4 "%08x"' "$1") ) +1 ))
        echo "white%red%green%blue%amber%violet%yellow%ir" | cut -d% -f "$val"
    }
 
    get_dt_name_from_color_function() {
        local ledpath="$1"
        local function
        local funcenum
        local color

        [ ! -f "$ledpath/color" ] && return 1

        color=$(get_dt_color_by_id "$ledpath/color")
        [ -z "$color" ] && return 1 # Something went wrong
        echo -n "$color"

        [ -f "$ledpath/function" ] && { 
                function=$(cat "$ledpath/function")
                [ -n "$function" ] && {
                        echo -n ":$function"

                        funcenum=$(cat "$ledpath/function-enumerator" 2>/dev/null)
                        [ -n "$funcenum" ] && echo -n "-$funcenum"
                }
        }
        return 0
    }

    get_dt_led_path() {
        local ledpath
        local basepath="/proc/device-tree"
        local nodepath="$basepath/aliases/led-$1"

        [ -f "$nodepath" ] && ledpath=$(cat "$nodepath")
        [ -n "$ledpath" ] && ledpath="$basepath$ledpath"

        echo "$ledpath"
    }

    get_dt_led() {
        local label
        local ledpath=$(get_dt_led_path $1)

        [ -n "$ledpath" ] && \
                label=$(cat "$ledpath/label" 2>/dev/null) || \
                label=$(cat "$ledpath/chan-name" 2>/dev/null) || \
                label=$(get_dt_name_from_color_function "$ledpath") || \
                label=$(basename "$ledpath")

        echo "$label"
    }

Reply:
> we can leave the basename there as a last resort. The led's common.yaml in 56 still states "If omitted, the label is taken from the node name [...]" and drivers like leds-pwm.c have not been converted. It's a barrel of fun.

Yes, I actually just moved it to the get_dt_led_name function in my draft, as I had some problems with the true/false in the chained OR command.

> There's another worse way to do it. This is the leds.sh portion. It uses a lookup table and needs to convert the color big-endian (because DT is big endian) value to a string.

I also thought about something like that; the problem is that there is not just color:function, function, color, function-number etc., but there might also be a device-name if the LED is linked to a parent device, and then you would have something like wlan0:green:activity for some leds. That was the point where I stopped to think about assembling the name myself :-(